Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1583,6 +1571,86 @@ public partial class ModuleResolver {
       }
     }
 
+    private void CheckForCyclesAmongRedirectingTypes(RedirectingTypeDecl dd, HashSet<ICallable> cycleErrorHasBeenReported) {
+      var enclosingModule = dd.EnclosingModule;
+      if (enclosingModule.CallGraph.GetSCCSize(dd) != 1) {
+        var r = enclosingModule.CallGraph.GetSCCRepresentative(dd);
+        if (cycleErrorHasBeenReported.Contains(r)) {
+          // An error has already been reported for this cycle, so don't report another.
+          // Note, the representative, "r", may itself not be a const.
+        } else if (dd is NewtypeDecl || dd is SubsetTypeDecl) {
```

Comment:
=> `is RedirectingType` ?

Reply:
`RedirectingType` also includes type synonyms, which has no constraints.

Reply:
Could you introducer an interface for the two? If I read `dd is NewtypeDecl || dd is SubsetTypeDecl` then I ask myself "Why those two?" and a named interface would answer that question.